### ISSUE TITLE
RetryFilter is ignored for IClusterClient.Connect if IGatewayListProvider.GetGateways() fails

### ISSUE THREAD

**Issue by pentp** [dotnet/orleans#6458]
If `GatewayManager` constructor fails to get the gateway list in 3 seconds, then DI fails and there is no way to retry the client connection attempt.
Generally it's considered a bad idea to do long running initialization inside constructors, so maybe some of the initialization code could be moved out and handled as part of the connection setup (and respect `retryFilter` parameter then also).

```
{TimeoutException("WithTimeout has timed out after 00:00:03") Source: 'Orleans.Core'
   at Orleans.Internal.OrleansTaskExtentions.<WithTimeout>d__19`1.MoveNext() in D:\build\agent\_work\25\s\src\Orleans.Core\Async\TaskExtensions.cs:line 286
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Orleans.AzureUtils.OrleansSiloInstanceManager.<FindAllGatewaySilos>d__21.MoveNext() in D:\build\agent\_work\25\s\src\Azure\Orleans.Clustering.AzureStorage\OrleansSiloInstanceManager.cs:line 149
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Orleans.AzureUtils.OrleansSiloInstanceManager.<FindAllGatewayProxyEndpoints>d__19.MoveNext() in D:\build\agent\_work\25\s\src\Azure\Orleans.Clustering.AzureStorage\OrleansSiloInstanceManager.cs:line 103
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Orleans.Messaging.GatewayManager..ctor(IOptions`1 gatewayOptions, IGatewayListProvider gatewayListProvider, ILoggerFactory loggerFactory, ConnectionManager connectionManager) in D:\build\agent\_work\25\s\src\Orleans.Core\Messaging\GatewayManager.cs:line 56
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ActivatorUtilities.ConstructorMatcher.CreateInstance(IServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ActivatorUtilities.CreateInstance(IServiceProvider provider, Type instanceType, Object[] parameters)
   at Microsoft.Extensions.DependencyInjection.ActivatorUtilities.CreateInstance[T](IServiceProvider provider, Object[] parameters)
   at Orleans.OutsideRuntimeClient.<StartInternal>d__56.MoveNext() in D:\build\agent\_work\25\s\src\Orleans.Core\Runtime\OutsideRuntimeClient.cs:line 236
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Orleans.OutsideRuntimeClient.<Start>d__55.MoveNext() in D:\build\agent\_work\25\s\src\Orleans.Core\Runtime\OutsideRuntimeClient.cs:line 208
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Orleans.ClusterClient.<Connect>d__20.MoveNext() in D:\build\agent\_work\25\s\src\Orleans.Core\Core\ClusterClient.cs:line 126
```

**Comment by pentp** [comment]
It looks like there is actually warmup code for this in `OutsideRuntimeClient.StartInternal`, but this doesn't guarantee that it will not fail on that exact call inside the constructor (and also the gateway list is queried twice during startup, if I read that correctly).

**Comment by ReubenBond** [comment]
Thanks for reporting, @pentp, I'm working on a fix

**Comment by ReubenBond** [comment]
Fix here #6458

We will release 3.1.5 with a fix soon (perhaps this week)
